core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java

```diff
@@ -605,31 +605,41 @@ public Object visitRemoveExpiredCommand(InvocationContext ctx, RemoveExpiredComm
 
    @Override
    public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throws Throwable {
-      long size = trySizeOptimization(command.getFlagsBitSet());
-      if (size > -1) {
+      long size = trySizeOptimization(ctx, command);
+      if (size >= 0) {
          return size;
       }
 
-      ClusterPublisherManager<?, ?> managerToUse;
-      if (command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-         managerToUse = localClusterPublisherManager;
-      } else {
-         managerToUse = clusterPublisherManager;
+      return asyncValue(localClusterPublisherManager.keyReduction(false, command.getSegments(), null,
+            ctx, command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(), PublisherReducers.add()));
+   }
+
+   private long trySizeOptimization(InvocationContext ctx, SizeCommand command) {
+      if (!command.hasAnyFlag(FlagBitSets.SKIP_SIZE_OPTIMIZATION)) {
+         // If the context is empty we can count the size for the given segments.
+         if (ctx == null || ctx.lookedUpEntriesCount() == 0) {
+            return dataContainer.size(identifySegments(command.getSegments()));
```

Comment:
This isn't actually an optimization (it requires iterating over the entire data container) and now that I think about this should just replace lines 613 and 614 above.

Reply:
You mean to use the `DataContainer#size(IntSet)` instead of the `ClusterPublisherManager#keyReduction`? If so, we would need to handle the keys in the context too.